cmd/upgrade/operatorroles/cmd.go

```diff
@@ -172,110 +280,166 @@ func run(cmd *cobra.Command, argv []string) {
 		if err != nil {
 			return err
 		}
-		policyARN, err = awsClient.EnsurePolicy(policyARN, string(policy),
-			aws.DefaultPolicyVersion, map[string]string{
-				tags.OpenShiftVersion: aws.DefaultPolicyVersion,
-				tags.RolePrefix:       prefix,
-				"operator_namespace":  operator.Namespace,
-				"operator_name":       operator.Name,
+		reporter.Debugf("Creating role '%s'", roleName)
+		roleARN, err := awsClient.EnsureRole(roleName, policy, "", "",
+			map[string]string{
+				tags.ClusterID:       cluster.ID(),
+				"operator_namespace": operator.Namespace,
+				"operator_name":      operator.Name,
 			})
 		if err != nil {
 			return err
 		}
-		reporter.Infof("Upgraded policy with ARN '%s' to version '%s'", policyARN, aws.DefaultPolicyVersion)
-		err = awsClient.UpdateTag(roleName)
+		reporter.Infof("Created role '%s' with ARN '%s'", roleName, roleARN)
+		reporter.Debugf("Attaching permission policy '%s' to role '%s'", policyARN, roleName)
+		err = awsClient.AttachRolePolicy(roleName, policyARN)
 		if err != nil {
-			return err
+			return fmt.Errorf("Failed to attach role policy. Check your prefix or run "+
+				"'rosa create account-roles' to create the necessary policies: %s", err)
 		}
 	}
 	return nil
 }
 
-func buildCommands(prefix string, accountID string, cluster *cmv1.Cluster, awsClient aws.Client) (string, error) {
-	commands := []string{}
-	for credrequest, operator := range aws.CredentialRequests {
-		roleName := aws.GetOperatorRoleName(cluster, operator)
-		iamRoleTags := fmt.Sprintf(
-			"Key=%s,Value=%s",
-			tags.OpenShiftVersion, aws.DefaultPolicyVersion)
-		tagRole := fmt.Sprintf("aws iam tag-role \\\n"+
-			"\t--tags %s \\\n"+
-			"\t--role-name %s",
-			iamRoleTags, roleName)
-		commands = append(commands, tagRole)
-
-		policyARN := aws.GetOperatorPolicyARN(accountID, prefix, operator.Namespace, operator.Name)
-		policyTags := fmt.Sprintf(
-			"Key=%s,Value=%s",
-			tags.OpenShiftVersion, aws.DefaultPolicyVersion,
-		)
-		isCompatible, err := awsClient.IsPolicyCompatible(policyARN, aws.DefaultPolicyVersion)
-		if err != nil {
-			return "", err
-		}
-		//We need because users might run it mutiple times and we dont want to create unnecessary versions
-		if isCompatible {
-			continue
-		}
-		createPolicy := fmt.Sprintf("aws iam create-policy-version \\\n"+
-			"\t--policy-arn %s \\\n"+
-			"\t--policy-document file://openshift_%s_policy.json \\\n"+
-			"\t --set-as-default",
-			policyARN, credrequest)
-		tagPolicy := fmt.Sprintf("aws iam tag-policy \\\n"+
-			"\t--tags %s \\\n"+
-			"\t--policy-arn %s",
-			policyTags, policyARN)
-
-		commands = append(commands, createPolicy, tagPolicy, tagRole)
+func getRoleName(cluster *cmv1.Cluster, missingOperator aws.Operator) string {
+	operatorIAMRoles := cluster.AWS().STS().OperatorIAMRoles()
+	rolePrefix := ""
+	for _, operatorIAMRole := range operatorIAMRoles {
+		roleName := strings.SplitN(operatorIAMRole.RoleARN(), "/", 2)[1]
```

Comment:
Maybe we should check here that the slice returned from `strings.SplitN` is greater than 1?
Though it should be, just for safety.